core/src/test/java/com/linecorp/armeria/client/InitiateConnectionShutdownTest.java

```diff
@@ -0,0 +1,229 @@
+import org.junit.jupiter.api.extension.RegisterExtension;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.CsvSource;
+
+import com.google.common.collect.Lists;
+
+import com.linecorp.armeria.common.AggregatedHttpResponse;
+import com.linecorp.armeria.common.HttpHeaderNames;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.HttpStatus;
+import com.linecorp.armeria.common.SessionProtocol;
+import com.linecorp.armeria.common.logging.RequestLogAccess;
+import com.linecorp.armeria.server.ServerBuilder;
+import com.linecorp.armeria.testing.junit5.server.ServerExtension;
+
+import io.netty.handler.codec.http.HttpHeaderValues;
+import io.netty.util.AttributeMap;
+
+public class InitiateConnectionShutdownTest {
+    @RegisterExtension
+    static final ServerExtension server = new ServerExtension() {
+        @Override
+        protected void configure(ServerBuilder sb) throws Exception {
+            sb.service("/ok", (ctx, req) -> HttpResponse.of(HttpStatus.OK));
+            sb.service("/delayed_ok", (ctx, req) -> {
+                return HttpResponse.delayed(HttpResponse.of(HttpStatus.OK),
+                                            Duration.ofMillis(500));
+            });
+        }
+    };
+
+    final List<RequestLogAccess> requestLogAccesses = Lists.newCopyOnWriteArrayList();
+    final AtomicBoolean completed = new AtomicBoolean();
+    final AtomicInteger connectionOpen = new AtomicInteger();
+    final AtomicInteger connectionClosed = new AtomicInteger();
+    final ConnectionPoolListener poolListener = new ConnectionPoolListener() {
+        @Override
+        public void connectionOpen(SessionProtocol protocol, InetSocketAddress remoteAddr,
+                                   InetSocketAddress localAddr, AttributeMap attrs) throws Exception {
+            connectionOpen.incrementAndGet();
+        }
+
+        @Override
+        public void connectionClosed(SessionProtocol protocol, InetSocketAddress remoteAddr,
+                                     InetSocketAddress localAddr, AttributeMap attrs) throws Exception {
+            connectionClosed.incrementAndGet();
+        }
+    };
+
+    @BeforeEach
+    void resetVariables() {
+        requestLogAccesses.clear();
+        completed.set(false);
+        connectionOpen.set(0);
+        connectionClosed.set(0);
+    }
+
+    @ParameterizedTest
+    @CsvSource({
+            "H1C, /ok,         BEFORE_SENDING_REQ",
+            "H1C, /ok,         AFTER_SENDING_REQ",
+            "H1C, /delayed_ok, BEFORE_SENDING_REQ",
+            "H1C, /delayed_ok, AFTER_SENDING_REQ",
+            "H2C, /ok,         BEFORE_SENDING_REQ",
+            "H2C, /ok,         AFTER_SENDING_REQ",
+            "H2C, /delayed_ok, BEFORE_SENDING_REQ",
+            "H2C, /delayed_ok, AFTER_SENDING_REQ"
+    })
+    void testConnectionShutdown(SessionProtocol protocol, String path, ConnectionShutdownTiming timing)
```

Comment:
Should we also test a case where an`HttpRequest` is abort before acquire a `Channel`. We can add a decorator that aborts a request with `HttpRequest.abort(...)`. `initiateConnectionShutdown()` should return a future that completes the exception.

Reply:
When I get a `HttpRequest` and call `HttpRequest.abort(...)`?
In the `decorator` scope, `HttpRequest.abort(...)` is not any working.
```
.decorator((delegate, ctx, req) -> {
    req.abort(new RuntimeException()); -> just run `whenComplete().complete(null);` of `EmptyFixedStreamMessage`
    return delegate.execute(ctx, req);
})
```

I tried to call `HttpResponse.abort(...)`, then I could catch `ch` is null, but `ex` is also null.

Reply:
https://github.com/line/armeria/pull/4708/commits/a3b441e790d361a39ab04f257310be65e018fd7d#diff-36ceaad17095d0cf501bef2c9fa2491673b4de6ea6bfc79ea2a4c1d2627763b5R76

Reply:
I'm not sure [this change](https://github.com/line/armeria/pull/4708/commits/a3b441e790d361a39ab04f257310be65e018fd7d#diff-36ceaad17095d0cf501bef2c9fa2491673b4de6ea6bfc79ea2a4c1d2627763b5R76) is correct.

Reply:
If a request fails at the following point, the decorator chains are executed but a connection hasn't been established. 
https://github.com/line/armeria/blob/80600030bad7317b9df55da2f89c349476a4cca5/core/src/main/java/com/linecorp/armeria/client/HttpClientDelegate.java#L64-L86

I found that it is difficult to simulate the error case using `HttpRequest.abort()`.
We can use the following hack, although it is not a normal case. 
```java
.decorator((delegate, ctx, req) -> {
    ClientPendingThrowableUtil.setPendingThrowable(ctx, new IllegalArgumentException());
    return delegate.execute(ctx, req);
})
```
Note that `RequestLog.whenRequestComplete()` never completes exceptionally.
https://github.com/line/armeria/blob/c7b1c84429d80b8bf437e538aea9be8db6d01ccc/core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java#L413-L429
What do you think of using `RequestLog.requestCause()` to get the cause?
```java
log().whenRequestComplete().thenAccept(log -> {
    if (log.channel() == null) {
         // Use log.requestCause() to exceptionally complete the future
    }
}

```

Reply:
I update test code with your help.